### PR TITLE
Image Patch Operator framework

### PR DESCRIPTION
**Pull request by muhilaru**
# Description

Created initial framework for Image Patch Operator and ImageBuildRequest custom resource. Configured make targets for make manifests, make generate, and make docker-build to reflect the structure used in the application and platform operator. Operator is built as part of the build pipeline and passes all compliance checks and unit tests. The ImageBuildRequest resource was successfully able to be created in a local Kubernetes cluster.

Unit tests will be added in a follow up PR.

Fixes VZ-2817

# Checklist 

As the author of this PR, I have:

- [x] Checked that I included or updated copyright and license notices in all files that I altered
- [ ] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: image-patch-operator/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Copyright (C) 2021, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+
+# Build the manager binary
+FROM golang:1.13 as builder
```
Comment (markxnelson) [inline image-patch-operator/Dockerfile:5]: you'll want to change this to use Oracle Linux, like the other operators

Comment (mgianatagh) [reply image-patch-operator/Dockerfile:5]: Yes, that is one of the next steps, just our first push to get the new sub-folder wired up

File: .gitignore [hunk 2]
```diff
@@ -39,5 +39,8 @@ out/
 **/*.log
 platform-operator/THIRD_PARTY_LICENSES.txt
 application-operator/THIRD_PARTY_LICENSES.txt
+image-patch-operator/THIRD_PARTY_LICENSES.txt
 tools/analysis/THIRD_PARTY_LICENSES.txt
 tests/e2e/**/*.test
+image-patch-operator/bin
```
Comment (bobdonat) [inline .gitignore:45]: Why are these needed?  I don't see them included for the other operators

File: image-patch-operator/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+WORKDIR /
+
+RUN groupadd -r verrazzano && useradd --no-log-init -r -g verrazzano -u 1000 verrazzano \
+    && mkdir /home/verrazzano \
+    && chown -R 1000:verrazzano /home/verrazzano
+
+COPY --from=builder --chown=verrazzano:verrazzano /usr/local/bin/verrazzano-image-patch-operator /usr/local/bin/verrazzano-image-patch-operator
+
+# Copy source tree to image
```
Comment (bobdonat) [inline image-patch-operator/Dockerfile:28]: This is not copying the source tree to image

File: image-patch-operator/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,33 @@
+# Copyright (C) 2021, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+
+FROM ghcr.io/oracle/oraclelinux:7-slim as builder
```
Comment (bobdonat) [inline image-patch-operator/Dockerfile:4]: Is a multi-stage build needed here?

File: image-patch-operator/PROJECT [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2021, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
+
+domain: verrazzano.io
+repo: github.com/verrazzano/verrazzano/image-patch-operator
+resources:
+- group: images
+  kind: ImageBuildRequest
+  version: v1alpha1
+version: "2"
```
Comment (bobdonat) [inline image-patch-operator/PROJECT:10]: version 2?

File: image-patch-operator/main.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+func main() {
+	var metricsAddr string
+	var enableLeaderElection bool
+	flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
+	flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
+		"Enable leader election for controller manager. "+
+			"Enabling this will ensure there is only one active controller manager.")
+	flag.Parse()
+
+	ctrl.SetLogger(zap.New(zap.UseDevMode(true)))
```
Comment (bobdonat) [inline image-patch-operator/main.go:42]: Probably a follow-on Jira. Use same logging as other operators.

File: image-patch-operator/Dockerfile (hunk 4 above)
Comment (markxnelson) [reply image-patch-operator/Dockerfile:4]: doesn't look like it is adding any benefit currently

**Comment by markxnelson** [comment]
if @bobdonat approves, i do too

File: image-patch-operator/Dockerfile (hunk 4 above)
Comment (mgianatagh) [reply image-patch-operator/Dockerfile:4]: this structure was copied from the application-operator Dockerfile.  If we don't need multi-stage here we can also eliminate a stage for the application-operator too

File: image-patch-operator/Dockerfile (hunk 3 above)
Comment (mgianatagh) [reply image-patch-operator/Dockerfile:28]: This is copied from the application-operator DockerFile, should be fixed in both places.  Also the platform-operator has the same comment.

File: .gitignore (hunk 2 above)
Comment (mgianatagh) [reply .gitignore:45]: I agree, these can be removed.  Probably a left over from when we initially created the kubebuiler project.

File: image-patch-operator/PROJECT (hunk 5 above)
Comment (muhilaru) [reply image-patch-operator/PROJECT:10]: Version 2 refers to the project version setup by Kubebuilder. We used the same project version as the application and platform operator.

File: image-patch-operator/main.go (hunk 6 above)
Comment (muhilaru) [reply image-patch-operator/main.go:42]: Jira for this task has been created.

File: .gitignore (hunk 2 above)
Comment (muhilaru) [reply .gitignore]: Removed.

File: image-patch-operator/Dockerfile (hunk 3 above)
Comment (muhilaru) [reply image-patch-operator/Dockerfile]: Removed this comment from all three operator dockerfiles.

**Review by desagar — APPROVED** [review]

**Review by bobdonat — APPROVED** [review]
